ledger/nomos-ledger/src/mantle/mod.rs

```diff
@@ -455,4 +628,106 @@ mod tests {
+            zk_sig: DummyZkSignature::prove(zksig::ZkSignaturePublic {
+                pks: vec![utxo.note.pk.into(), declare_op.zk_id.0],
+                msg_hash: declare_tx_hash.0,
+            }),
+            ed25519_sig: signing_key.sign(declare_tx_hash.as_signing_bytes().as_ref()),
+        })];
+
+        let ledger_state = ledger_state
+            .try_apply_tx::<MainnetGasConstants>(
+                0,
+                &test_config,
+                cryptarchia_state.latest_commitments(),
+                declare_tx,
+            )
+            .unwrap();
+
+        assert!(ledger_state.sdp.get_declaration(&declaration_id).is_ok());
+        assert!(ledger_state
+            .locked_notes
+            .contains(&declare_op.locked_note_id));
+
+        // Apply the active operation.
+        let active_op = SDPActiveOp {
+            declaration_id,
+            nonce: 1,
+            metadata: None,
+        };
+        let mut active_tx = create_test_tx_with_ops(vec![Op::SDPActive(active_op)]);
+        let active_tx_hash = active_tx.hash();
+        active_tx.ops_proofs = vec![Some(OpProof::ZkSig(DummyZkSignature::prove(
+            zksig::ZkSignaturePublic {
+                pks: vec![utxo.note.pk.into(), declare_op.zk_id.0],
+                msg_hash: active_tx_hash.0,
+            },
+        )))];
+
+        let ledger_state = ledger_state
+            .try_apply_tx::<MainnetGasConstants>(
+                1,
+                &test_config,
+                cryptarchia_state.latest_commitments(),
+                active_tx,
+            )
+            .unwrap();
+
+        let declaration = ledger_state.sdp.get_declaration(&declaration_id).unwrap();
+        assert_eq!(declaration.active, 1);
+        assert_eq!(declaration.nonce, 1);
+
+        // Apply the withdraw operation.
+        let withdraw_op = SDPWithdrawOp {
+            declaration_id,
+            nonce: 2,
+        };
+        let mut withdraw_tx = create_test_tx_with_ops(vec![Op::SDPWithdraw(withdraw_op)]);
+        let withdraw_tx_hash = withdraw_tx.hash();
+        withdraw_tx.ops_proofs = vec![Some(OpProof::ZkSig(DummyZkSignature::prove(
+            zksig::ZkSignaturePublic {
+                pks: vec![utxo.note.pk.into(), declare_op.zk_id.0],
+                msg_hash: withdraw_tx_hash.0,
+            },
+        )))];
+        let ledger_state = ledger_state
+            .try_apply_tx::<MainnetGasConstants>(
+                11,
+                &test_config,
+                cryptarchia_state.latest_commitments(),
+                withdraw_tx,
+            )
+            .unwrap();
+
+        let declaration = ledger_state.sdp.get_declaration(&declaration_id).unwrap();
+        assert!(!ledger_state
+            .locked_notes
+            .contains(&declare_op.locked_note_id));
+        assert_eq!(declaration.active, 1);
+        assert_eq!(declaration.withdrawn, Some(11));
+        assert_eq!(declaration.nonce, 2);
+    }
```

Comment:
is there a test to check early withdrawal? can we add it?

Reply:
There is this test in sdp state transition tests: https://github.com/logos-co/nomos/blob/master/nomos-core/chain-defs/src/sdp/state.rs#L441-L460

Do you think I should add one in the `ledger/nomos-ledger/src/mantle/mod.rs`?